janus_core/src/message.rs

```diff
@@ -1093,96 +1149,390 @@ impl Decode for Report {
+    ///
+    /// This method would typically be used for code which is generic over the query type.
+    /// Query-type specific code will typically call [`Self::batch_id`].
+    pub fn batch_identifier(&self) -> &Q::CollectRespBatchIdentifier {
+        &self.batch_identifier
+    }
+
+    /// Gets the report count associated with this collect response.
+    pub fn report_count(&self) -> u64 {
+        self.report_count
+    }
+
+    /// Gets the encrypted aggregate shares associated with this collect response.
+    pub fn encrypted_aggregate_shares(&self) -> &[HpkeCiphertext] {
+        &self.encrypted_aggregate_shares
+    }
+}
+
+impl CollectResp<TimeInterval> {
+    /// Constructs a new collect response for a time-interval task.
+    pub fn new_time_interval(
+        report_count: u64,
+        encrypted_aggregate_shares: Vec<HpkeCiphertext>,
+    ) -> Self {
+        Self::new((), report_count, encrypted_aggregate_shares)
+    }
+}
+
+impl CollectResp<FixedSize> {
+    /// Constructs a new collect response for a fixed-size task.
+    pub fn new_fixed_size(
+        batch_id: BatchId,
+        report_count: u64,
+        encrypted_aggregate_shares: Vec<HpkeCiphertext>,
+    ) -> Self {
+        Self::new(batch_id, report_count, encrypted_aggregate_shares)
+    }
+
+    // Gets the batch ID associated with this collect response.
+    pub fn batch_id(&self) -> &BatchId {
+        self.batch_identifier()
+    }
 }
 
-impl Encode for CollectResp {
+impl<Q: QueryType> Encode for CollectResp<Q> {
     fn encode(&self, bytes: &mut Vec<u8>) {
-        encode_u16_items(bytes, &(), &self.encrypted_agg_shares);
+        Q::CODE.encode(bytes);
+        self.batch_identifier.encode(bytes);
+        self.report_count.encode(bytes);
+        encode_u16_items(bytes, &(), &self.encrypted_aggregate_shares); // TODO(#471): should be encode_u32_items
     }
 }
 
-impl Decode for CollectResp {
+impl<Q: QueryType> Decode for CollectResp<Q> {
     fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
-        let encrypted_agg_shares = decode_u16_items(&(), bytes)?;
+        query_type::Code::decode_expecting_value(bytes, Q::CODE)?;
+        let batch_identifier = Q::CollectRespBatchIdentifier::decode(bytes)?;
+        let report_count = u64::decode(bytes)?;
+        let encrypted_aggregate_shares = decode_u16_items(&(), bytes)?; // TODO(#471): should be decode_u32_items
 
         Ok(Self {
-            encrypted_agg_shares,
+            batch_identifier,
+            report_count,
+            encrypted_aggregate_shares,
         })
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::{
-        CollectReq, CollectResp, Duration, Extension, ExtensionType, HpkeAeadId, HpkeCiphertext,
-        HpkeConfig, HpkeConfigId, HpkeKdfId, HpkeKemId, HpkePublicKey, Interval, Nonce, Report,
-        ReportMetadata, Role, TaskId, Time,
-    };
-    use assert_matches::assert_matches;
+pub mod query_type {
```

Comment:
nit/thought: If this is going to be a nested module, we could move it to its own file (`janus_core/src/message/query_type.rs`) just to limit the size of this file. I'm not sure if I care about how big any individual source file gets, though.

Reply:
Hmm -- I don't have strong feelings here about which way is best. My notion would be to leave this as-is, on the thought that a couple dozen extra lines of code in this file will probably be easier to navigate that an additional file in our codebase. But I'm happy to move if folks think that's the better course?

Reply:
I don't have a strong opinion, so let's leave it as-is.